Improve ResupplyAircraft

reaperrr opened the pull request:
This resolves two dependencies to make `ResupplyAircraft` fully compatible with use as ChildActivity (particularly in `ReturnToBase`).


1) Replaced the work-around for #15055, as it causes complications in the above use-case.
`ResupplyAircraft` is now immediately queued again from `OnBecomingIdle` if the actor hasn't finished rearming/repairing yet.


2) On bleed, `ResupplyAircraft` (or rather its child activity `WaitFor`) continues to run until the actor is given an order or forced to yield reservation.
This makes it hard to use it as-is as ChildActivity (for example in ReturnToBase), because it makes its parent implicitly continue until the same conditions are met.
By instead preventing auto-TakeOff in Aircraft.OnBecomingIdle when `TakeOffOnResupply` is set to `false`, `ResupplyAircraft` itself no longer needs to take care of that, allowing it to finish properly as soon as the actor finishes resupply.

Split from #16241 and therefore dependency for it.

reaperrr committed:
Simplify ResupplyAircraft

By moving part of the take-off prevention (when TakeOffOnResupply
is set to false) to Aircraft.

Main reason & advantage is that dropping the 'WaitFor' child
makes this activity always end when resupplies are done,
which makes it more compatible with being queued as ChildActivity
itself (for example by ReturnToBase).

tovl commented:
Works as advertised. Just an idea though: I'm thinking that maybe simply letting `OnBecomingIdle` queue `ResupplyAircraft` when the aircraft is positioned on an appropriate resupply building would solve #15055 in just a few extra lines.

reaperrr committed:
Make Aircraft.GetActorBelow() LandAltitude-aware

matjaeck commented:
Both issues that I mentioned in #16241 are fixed here. This also changes the behavior described in #16320: Airplanes will now always continue their attack-move after reloading, regardless of the number of airfields. When they do not find a target, they RTB but immediately take off again and continue their attack-move. See https://imgur.com/a/9opZWcJ.

reaperrr commented:
> Both issues that I mentioned in #16241 are fixed here.

The issues were caused by the changes in `ReturnToBase`, not in `ResupplyAircraft`, so they're not expected to show up here.

> When they do not find a target, they RTB but immediately take off again and continue their attack-move.

As long as `AbortOnResupply` is set to `false`, that's basically the expected behaviour, I think. At least changing that behavior further is out of scope for this PR.

reaperrr committed:
Fix ResupplyAircraft being cancelable by Stop command

It is now immediately queued again, as long as the actor
has not finished rearming/repairing yet.

reaperrr commented:
Updated.

> Just an idea though: I'm thinking that maybe simply letting OnBecomingIdle queue ResupplyAircraft when the aircraft is positioned on an appropriate resupply building would solve #15055 in just a few extra lines.

Done, though it took more than just a few extra lines to (hopefully) get it right...

matjaeck commented:
> The issues were caused by the changes in ReturnToBase, not in ResupplyAircraft, so they're not expected to show up here.

Stupid me.

> As long as AbortOnResupply is set to false, that's basically the expected behaviour, I think. At least changing that behavior further is out of scope for this PR.

No idea, it was just something I noticed.

obrakmann commented on OpenRA.Game/Activities/Activity.cs:
```diff
@@ -144,15 +144,19 @@ public Activity()
 		}
 
+		// HACK: Ensure that OnFirstRun always triggers before the first Tick(self), even if Actor.CurrentActivity was updated in a 'dirty' way
+		// that would otherwise skip OnFirstRun (for example through Actor.QueueActivity).
```
Wait, what? How does `Actor.QueueActivity` cause skipping `OnFirstRun`?

reaperrr replied:
Well, maybe I misinterpreted the cause, but
```
self.CancelActivity();
self.QueueActivity(new ResupplyAircraft(self));
```
results in `OnFirstRun` not triggering (it seems to go right to `Tick` instead, at least before this commit).

obrakmann replied:
So I just tested on bleed, and I can't reproduce:

```
Internal mods:
	all: All mods (git-16f1750252)
	cnc: Tiberian Dawn (git-16f1750252)
	d2k: Dune 2000 (git-16f1750252)
	modcontent: Mod Content Manager (git-16f1750252)
	ra: Red Alert (git-16f1750252)
	ts: Tiberian Sun (git-16f1750252)
External mods:
	cnc-git-fd5eaf1f49: Tiberian Dawn (git-fd5eaf1f49)
	d2k-git-09512f0987: Dune 2000 (git-09512f0987)
	ra-playtest-20180729: Red Alert (playtest-20180729)
	ts-git-c1b9df439e: Tiberian Sun (git-c1b9df439e)
	ra-git-16f1750252: Red Alert (git-16f1750252)
Loading mod: ra
System.Net.Sockets.SocketOptionName 0x17 is not supported at IP level
[2019-03-22T21:06:48] Game started
ResupplyAircraft.OnFirstRun()
ResupplyAircraft.Tick()
ResupplyAircraft.Tick()
ResupplyAircraft.Tick()
```

Do I have to follow any special sequence of events, or is ordering an aircraft to an airfield enough?

reaperrr replied:
It may not be reproducable on bleed because on bleed, it uses 'normal' queueing instead of ChildActivity, so unless you give the order when already on the pad/field, it has long finished before the aircraft arrives to reload.

To reproduce it on this PR: Revert the last two commits, compile, then in RA let some aircraft deplete some of its ammo, order it to resupply, and as soon as it touches down on the pad/afld and you see the resupply anim starting, press Stop, so the activity is canceled. Now press Stop again, and `ResupplyAircraft.OnFirstRun()` should not trigger, even through the activity was queued after CancelActivity.

tovl replied:
This is really weird. I poked around a bit and found that `ResupplyActivity` for some arcane reason goes into the `canceling` state before it even gets the chance to tick for the first time. However, the moment the activity is queued it _is_ in the `queued` state as it should. This happens even when there is no previous activity. So somehow something is doing weird things with the activity state somewhere in between those moments.

obrakmann replied:
OK, that was stupid.  The reason why that happens is because the Stop order is handled by every trait under the sun, not only `Aircraft`.  In this case, at least `AttackBase` also gets a shot and promptly cancels the activity that was just queued.

The fix would have been to queue the activity in a FrameEndTask.

pchote replied:
Wait, what? Why is `AttackBase` allowed to cancel non-attack orders?
Fixing that would be trickier, but IMO a better solution using a FrameEndTask workaround.

tovl replied:
Interesting. I agree that fixing `AttackBase` would be the best solution in the long run, but I don't think it is reasonable to expect this PR to do that. I think the various attack activities need a refactor anyway (e.g. `Attack` is quite misleadingly named IMO).

In the context of what this PR is trying to do, it is not needed. Because the last commit makes the requeueing of `ResupplyActivity` happen on becoming idle, it isn't affected by this bug anymore. You can just drop the second to last commit and everything still works fine.

obrakmann replied:
> Why is AttackBase allowed to cancel non-attack orders?

AFAIK we don't check the type of `Actor.CurrentActivity` anywhere where the Stop order gets resolved, we just call `Actor.CancelActivity` without regard to that (which is why it gets called multiple times every time you press S, once from each applicable `ResolveOrder` method). It's only never - until recently - been a problem because we never automatically queued new activities on Stop.

But yes, as tovl mentioned, the FrameEndTask fix is hypothetical since the last commit removes the offending code.  But the second-to-last commit should be removed.

reaperrr replied:
> You can just drop the second to last commit and everything still works fine.

Will do that.

About a future proper fix, I think there's several activities we don't want to cancel on "Stop", but we do want them to cancel on pretty much anything else, so maybe we should chain a `bool cancelAttemptByStopOrder` as optional parameter from CancelActivity all the way through to Activity, to allow activities to refuse cancellation by "Stop" orders.

We could also pass through the order string to make things more configurable, but that would be a bit more complicated, and I'm not sure we really need that.

obrakmann replied:
> About a future proper fix, I think there's several activities we don't want to cancel on "Stop", but we do want them to cancel on pretty much anything else

What other examples are you thinking of besides `ResupplyAircraft`?

I think the issue Paul is seeing is that traits were stopping an activity unrelated to that trait, not that the activity could be stopped in the first place.

pchote replied:
> I think the issue Paul is seeing is that traits were stopping an activity unrelated to that trait, not that the activity could be stopped in the first place.

Yeah, exactly.

Re customizing activity stop behaviour: we could implement a `Stop` method on `Activity` with a default implementation that cancels the activity. Special cases can then override this to do other things.

reaperrr replied:
> What other examples are you thinking of besides ResupplyAircraft?

`Resupply` (once that PR is merged) for ground units and [Heli]Land, at least. There may be one or two more, but these came to mind first.

obrakmann commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -531,22 +530,52 @@ void INotifyBecomingIdle.OnBecomingIdle(Actor self)
+			else if (atLandAltitude && Info.CanHover && Info.TakeOffOnResupply)
+			{
+				self.QueueActivity(new TakeOff(self));
+				UnReserve();
```
`TakeOff` will run `UnReserve` itself.

obrakmann commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -531,22 +530,52 @@ void INotifyBecomingIdle.OnBecomingIdle(Actor self)
-			else if (!Info.CanHover && (Info.TakeOffOnResupply || ReservedActor == null || self.World.Map.DistanceAboveTerrain(CenterPosition) != Info.LandAltitude))
+			else if (!Info.CanHover && (Info.TakeOffOnResupply || ReservedActor == null || !atLandAltitude))
+			{
 				self.QueueActivity(new FlyCircle(self, -1, Info.IdleTurnSpeed > -1 ? Info.IdleTurnSpeed : TurnSpeed));
```
I may be missing something, but this would skip the TakeOff activity, wouldn't it?

reaperrr replied:
Well, aircraft still take off automatically when a Fly activity triggers, but I guess you're right, it would be better for consistency to queue a TakeOff instead.

reaperrr commented:
Rebased & updated.

reaperrr commented:
Updated.

obrakmann commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -521,20 +530,44 @@ public void ModifyDeathActorInit(Actor self, TypeDictionary init)
+				// Temporary HACK for the AutoCarryall special case (needs CanHover, but also HeliFlyCircle on idle).
+				// Will go away soon (in a separate PR) with the arrival of ActionsWhenIdle.
 				self.QueueActivity(new HeliFlyCircle(self, Info.IdleTurnSpeed > -1 ? Info.IdleTurnSpeed : TurnSpeed));
+			}
```
I'm really looking forward to #15873, this whole chain of conditionals is so horrible :rofl:

reaperrr replied:
Yeah. I'm actually thinking about pulling in #15803 first because less activities + properly untangled CanHover/VTOL would make #15873 a bit simpler than it would currently be.

obrakmann reviewed (changes requested):
Not sure how we missed that crash the first time around.

reaperrr commented on OpenRA.Mods.Common/Activities/Air/ResupplyAircraft.cs:
```diff
@@ -28,18 +28,10 @@ protected override void OnFirstRun(Actor self)
-			{
-				ChildActivity = ActivityUtils.SequenceActivities(self, aircraft.GetResupplyActivities(host).ToArray());
-				QueueChild(self, new AllowYieldingReservation(self));
+			ChildActivity = ActivityUtils.SequenceActivities(self, aircraft.GetResupplyActivities(host).ToArray());
```
Hm, ok. This should already be an issue on bleed and would be fixed by #16352 anyway, but I guess I can throw in a temporary fix.

obrakmann replied:
No, bleed is fine. The reason this crashes is that you've changed `GetResupplyActivities` to evaluate whether anything needs to be done. Bleed just returns all the activities.

reaperrr replied:
Ah, I see. Well, should be fixed now.

reaperrr commented:
Updated, crash should be fixed now.

reaperrr commented:
Updated.

obrakmann reviewed (dismissed):
Looks fine to me now, :+1:

tovl reviewed (dismissed):
This fixes #15055 and found no regressions. :+1:

reaperrr commented:
Rebased.

reaperrr commented:
I count 2 approvals, so lgtm.

reaperrr commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/77fe4b9ad76d72182536e439ddc40a93b8c4fc6a...f15a4f9f85212afe6442656b0e74e73494cd92cd
